apiserver/facades/client/charms/client.go

```diff
@@ -380,6 +412,77 @@ func (a *API) addCharmWithAuthorization(args params.AddCharmWithAuth) (params.Ch
 	}, nil
 }
 
+func (a *API) queueAsyncCharmDownload(args params.AddCharmWithAuth) (corecharm.Origin, error) {
+	charmURL, err := charm.ParseURL(args.URL)
+	if err != nil {
+		return corecharm.Origin{}, err
+	}
+
+	// Check if a charm doc already exists for this charm URL. If so, the
+	// charm has already been queued for download so this is a no-op. We
+	// still need to return back the recorded origin from where the charm
+	// is to be downloaded.
+	if _, err := a.backendState.Charm(charmURL); err == nil {
+		// Try to lookup the origin associated with the application.
+		// Due to the way that charm deploys are implemented by the
+		// client, if the actual deploy call fails, the resolved origin
+		// returned by this method will not be recorded against the
+		// application (as the application doc will not be created).
+		//
+		// If we are unable to figure out the origin, we still need
+		// to execute the remainder of this method
```

Comment:
You're in a weird state if the backend state has the charm, but doesn't have the uploaded charm origin. Are we not opening a race here?

Reply:
The way it works ATM is:
- AddCharm X -> returns ResolvedOrigin
- Deploy X (app) + ResolvedOrigin -> records origin to app document

Because this is not transactional, if the Deploy call fails (e.g. intermittent network issues) you have the charm document but no origin. So if you retry the `juju deploy X` command, we still have to fetch the metadata and re-calculate the resolved origin (which might in an edge case change between running the two `juju deploy` commands) so the `Deploy` can work as expected.

Note that in the future, all these things will happen in a single transaction (and in a dedicated facade) so I don't anticipate any issues. For now, this approach allows us to test the feature without accidentally breaking the existing deploy flow.